pyat/at/tracking/particles.py

```diff
@@ -0,0 +1,131 @@
+"""
+Functions relating to particle generation
+"""
+import numpy
+from at.physics import ohmi_envelope
+
+__all__ = ['beam', 'sigma_matrix']
+
+
+def sigma_matrix(argin, twiss=False):
+    """
+    Calculate the correlation matrix to be used for particle generation
+
+    PARAMETERS
+        argin           Lattice object or list of twiss parameters.
+                        Depending on the length of the list the function
+                        will do different things:
+                        [espread, blength] 
+                            computes 2x2 longitudinal sigma matrix
+                        [beta, alpha, emit] 
+                            computes 2x2 transverse sigma matrix
+                        [betax, alphax, emitx, betay, alphay, emity] 
+                            computes the 4x4 sigma matrix for horizontal 
+                            and vertical.
+                        [betax, alphax, emitx, betay, alphay, emity, espread, 
+                            blength] 
+                            computes the 6x6 sigma matrix.
+
+    KEYWORDS
+        twiss=False     Flag which states whether the input is a
+                        lattice object or list of twiss parameters
+
+    OUTPUT
+        sigma_matrix    correlation matrix (either 2x2, 4x4 or 6x6)
+
+    """
+    if not twiss:
+        cflag = False
+        if not argin.radiation:
+            argin.radiation_on()
+            cflag = True
+
+        emit0, beamdata, emit = ohmi_envelope(argin, refpts=[0])
+        sig_matrix = emit.r66[0]
+
+        if cflag:
+            argin.radiation_off()
+
+
+    else:
+        if len(argin) == 2:
+            [espread, blength] = argin
+            sig_matrix = numpy.array([
+                            [espread*espread, 0],
+                            [0, blength*blength]
+                                     ])
+
+        elif len(argin) == 3:
+            [bx, ax, epsx] = argin
+            sig_matrix = epsx * numpy.array([
+                                    [bx, -ax],
+                                    [-ax, (1 + ax * ax)/bx]
+                                            ])
+
+        elif len(argin) == 6:
+            [bx, ax, epsx, by, ay, epsy] = argin
+            sig_matrix = numpy.bmat([
```

Comment:
Rather than `numpy.bmat` which return a matrix object, `numpy.block` returns a standard nd-array, like emit.r66 in the first case.

Reply:
I think numpy.block was the function I was looking for. I had some problems with exactly this function that was returning a matrix object. numpy.block works in the same way but return a nd-array so it's perfect.